### PR TITLE
GH-43228: [C++] Fix Abseil compile error on GCC 13

### PR DESCRIPTION
**Pull request by amol-**
### Rationale for this change

When trying to compile Arrow with GCC 13, it fails due to ABSEIL missing a `<cstdint>` include, this PR addresses the issue by adding the missing include.

There have been past reports for this issue too: https://github.com/apache/arrow/issues/36969 

This is a more minimal fix that tries to avoid the complexity of previous attempts like  https://github.com/apache/arrow/pull/43147 and https://github.com/apache/arrow/pull/37066 which involved updating Abseil and facing additional issues to fix.

### What changes are included in this PR?

Add the missing include when GCC>=13

### Are these changes tested?

They are tested by the existing compile infrastructure and testsuite and by adding a new GCC-13 based CPP test environment for bundled builds.

### Are there any user-facing changes?

No, all behaviours should remain the same
* GitHub Issue: #43228

### CODE REVIEW

**Comment by amol-** [comment]
@github-actions crossbow submit -g cpp -g r

**Comment by amol-** [comment]
@raulcd @assignUser this should be ready for review, I tried a more conservative approach as upgrading Abseil introduced more problems than it fixed even on the most conservative version upgrade.

**Comment by kou** [comment]
Could you open a new issue for this instead of reusing existing closed issue?

**Comment by amol-** [comment]
> Could you open a new issue for this instead of reusing existing closed issue?

Given it's the same exact error message shouldn't we preserve a consolidated history by keeping the original issue?
Happy to open a new issue if that's the preference

**Comment by raulcd** [comment]
The problem is that the original issue is in the `13.0.0` milestone and all our current release scripts take that into account to generate release notes, etcetera. If we reuse the same issue it would be as if this PR was released on `13.0.0` and if we require to back port this issue it gets messy.
It has it's pros and cons but the current process doesn't account for two PRs pointing to the same GH issue.

**Comment by amol-** [comment]
> The problem is that the original issue is in the `13.0.0` milestone 

Makes sense, I didn't check and thought it was closed without being done. I'll open a new one.

File: cpp/cmake_modules/ThirdpartyToolchain.cmake [hunk 1]
```diff
@@ -2908,6 +2908,9 @@ macro(build_absl)
   set(ABSL_INCLUDE_DIR "${ABSL_PREFIX}/include")
   set(ABSL_CMAKE_ARGS "${EP_COMMON_CMAKE_ARGS}" -DABSL_RUN_TESTS=OFF
                       "-DCMAKE_INSTALL_PREFIX=${ABSL_PREFIX}")
+  if(CMAKE_COMPILER_IS_GNUCC AND CMAKE_CXX_COMPILER_VERSION VERSION_GREATER_EQUAL 13.0)
+    set(ABSL_CMAKE_ARGS ${ABSL_CMAKE_ARGS} "-DCMAKE_CXX_FLAGS=-include stdint.h")
```
Comment (kou) [inline cpp/cmake_modules/ThirdpartyToolchain.cmake:2912]:
```suggestion
    list(APPEND ABSL_CMAKE_ARGS "-DCMAKE_CXX_FLAGS=-include stdint.h")
```

**Review by kou — COMMENTED** [review]
Can we reproduce this by the following change?

```diff
diff --git a/dev/tasks/tasks.yml b/dev/tasks/tasks.yml
index 45417acf85..cd5f119b77 100644
--- a/dev/tasks/tasks.yml
+++ b/dev/tasks/tasks.yml
@@ -1418,6 +1418,16 @@ tasks:
         GCC_VERSION: 12
       image: ubuntu-r-only-r
 
+  # This also has -flto=auto
+  test-r-gcc-13:
+    ci: github
+    template: docker-tests/github.linux.yml
+    params:
+      env:
+        UBUNTU: 24.04
+        GCC_VERSION: 13
+      image: ubuntu-r-only-r
+
   test-r-minimal-build:
     ci: azure
     template: r/azure.linux.yml
```

File: cpp/cmake_modules/ThirdpartyToolchain.cmake (hunk 1 above)
Comment (amol-) [reply cpp/cmake_modules/ThirdpartyToolchain.cmake:2912]: Doesn't `list(APPEND` add an extra `;`?

Comment (kou) [reply cpp/cmake_modules/ThirdpartyToolchain.cmake:2912]: `set(X ${X} Y)` and `list(APPEND X Y)` do the same thing.

**Comment by amol-** [comment]
@github-actions crossbow submit test-r-gcc-13

**Comment by amol-** [comment]
@github-actions crossbow submit test-ubuntu-24.04-cpp-gcc-13

**Comment by amol-** [comment]
@github-actions crossbow submit test-ubuntu-24.04-cpp-gcc-13

**Comment by amol-** [comment]
@github-actions crossbow submit test-ubuntu-24.04-cpp-gcc-13

**Comment by amol-** [comment]
@github-actions crossbow submit test-ubuntu-24.04-cpp-gcc-13

**Comment by amol-** [comment]
@github-actions crossbow submit test-ubuntu-24.04-cpp-gcc-13-bundled

**Comment by amol-** [comment]
@github-actions crossbow submit test-ubuntu-24.04-cpp-gcc-13-bundled

**Comment by amol-** [comment]
@github-actions crossbow submit test-ubuntu-24.04-cpp-gcc-13-bundled

**Comment by amol-** [comment]
@github-actions crossbow submit test-ubuntu-24.04-cpp-gcc-13-bundled

**Comment by kou** [comment]
Hmm. Could you also add `-e ARROW_VERBOSE_THIRDPARTY_BUILD=ON`?

**Comment by amol-** [comment]
@github-actions crossbow submit test-ubuntu-24.04-cpp-gcc-13-bundled

**Comment by amol-** [comment]
@github-actions crossbow submit test-ubuntu-24.04-cpp-gcc-13-bundled

**Comment by amol-** [comment]
@github-actions crossbow submit test-ubuntu-24.04-cpp-gcc-13-bundled

**Comment by amol-** [comment]
@kou `test-ubuntu-24.04-cpp-gcc-13-bundled` now compiles correctly. It fails the test, but it seems unrelated as they are timezone related tests. I think that this is ready for final review.

**Comment by amol-** [comment]
@github-actions crossbow submit test-ubuntu-20.04-cpp-bundled

**Comment by amol-** [comment]
@github-actions crossbow submit test-ubuntu-24.04-cpp-gcc-13-bundled

**Comment by amol-** [comment]
The timestamp tests now pass too by installing `tzdata-legacy`

**Review by assignUser — APPROVED** [review]
Looks good, thanks! Could you also remove the workaround we added to r/configure?

**Comment by amol-** [comment]
@github-actions crossbow submit test-r-library-r-base-latest test-r-depsource-bundled

**Comment by amol-** [comment]
@github-actions crossbow submit test-r-depsource-bundled

**Comment by amol-** [comment]
> Looks good, thanks! Could you also remove the workaround we added to r/configure?

Done and tested on `test-r-depsource-bundled`

**Comment by assignUser** [comment]
@github-actions crossbow submit r-binary-packages 
Nice, thanks. I'll run this as it has a build on 24.04 which comes with the new gcc. I'll merge after

**Comment by assignUser** [comment]
Ugh, turns out `ubuntu-latest` is still 22.04, I modified the crossbow workflow to re-run on 24.04

**Pull request merged**